apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java

```diff
@@ -64,25 +64,32 @@ public void testUpdateBranchGrayRulesWithUpdateOnce() {
 
     namespaceBranchService.updateBranchGrayRules(testApp, testCluster, testNamespace, testBranchName, rule);
 
-    GrayReleaseRule
-        activeRule =
-        namespaceBranchService.findBranchGrayRules(testApp, testCluster, testNamespace, testBranchName);
+    GrayReleaseRule activeRule = namespaceBranchService.findBranchGrayRules(testApp, testCluster, testNamespace,
+        testBranchName);
 
     Assert.assertNotNull(activeRule);
     Assert.assertEquals(rule.getAppId(), activeRule.getAppId());
     Assert.assertEquals(rule.getRules(), activeRule.getRules());
     Assert.assertEquals(Long.valueOf(0), activeRule.getReleaseId());
 
-    Page<ReleaseHistory> releaseHistories = releaseHistoryService.findReleaseHistoriesByNamespace
-        (testApp, testCluster, testNamespace, pageable);
+    Page<ReleaseHistory> releaseHistories = releaseHistoryService.findReleaseHistoriesByNamespace(testApp, testCluster,
+        testNamespace, pageable);
 
     ReleaseHistory releaseHistory = releaseHistories.getContent().get(0);
 
     Assert.assertEquals(1, releaseHistories.getTotalElements());
     Assert.assertEquals(ReleaseOperation.APPLY_GRAY_RULES, releaseHistory.getOperation());
     Assert.assertEquals(0, releaseHistory.getReleaseId());
     Assert.assertEquals(0, releaseHistory.getPreviousReleaseId());
-    Assert.assertTrue(releaseHistory.getOperationContext().contains(rule.getRules()));
+    Assert.assertTrue(containRule(releaseHistory.getOperationContext(), rule.getRules()));
```

Comment:
I understand the original `contains` check is not a good one, however, I would like to know the background behind this change. Is this test failed in some situations?
BTW, I would suggest parsing the string to object for comparison instead of simply comparing the string, as there is not much difference from the original check.

Reply:
> I understand the original `contains` check is not a good one, however, I would like to know the background behind this change. Is this test failed in some situations? BTW, I would suggest parsing the string to object for comparison instead of simply comparing the string, as there is not much difference from the original check.

* Is this test failed in some situations?

  Yes, the test failed in certain cases.  I detected that the test results are flaky using the NonDex tool for catching non-deterministic/flaky tests (https://github.com/TestingResearchIllinois/NonDex).

  You may reproduce the nondeterministic behavior using the following steps:
  Clone the repo and cd into it.
  `mvn install -am -pl RosettaCore -DskipTests`
  `mvn -pl RosettaCore test -Dtest=com.hubspot.rosetta.RosettaBinderTest#itHandlesAnnotatedOptionalGenericSetterSerialization - Confirm that the test passes as is.`
  `mvn -pl RosettaCore edu.illinois:nondex-maven-plugin:1.1.2:nondex -DnondexMode=ONE -Dtest=com.hubspot.rosetta.RosettaBinderTest#itBindsStoredAsJsonBeanCorrectly` - This will run the NonDex tool on the test. 

* Parsing the string to object for comparison instead of simply comparing the string

  Thanks for pointing this out! I have modified the code and currently, I am using [GrayReleaseRuleItemTransformer#batchTransformFromJSON](https://github.com/apolloconfig/apollo/blob/301d66dcc28b2d698eefed3ecc9aa3273afb9ddf/apollo-common/src/main/java/com/ctrip/framework/apollo/common/utils/GrayReleaseRuleItemTransformer.java#L35) to transfer the string to object and compare the element within the set. Let me know if you have any suggestions on this :)